switch to std::simd, expand SIMD & docs

TheIronBorn opened the pull request:
Closes #1232. Should also fix #1162.

The new API is pretty easy and the new trait system made for some convenient generic implementations.

Changes:
* moved `__m128i` & `__m256i` away from `simd_support` feature
* added `__m512i` and some internal AVX512 optimizations
* dropped `u8x2`/`i8x2` types due to lack of support in `std::simd`
* added SIMD types (and `NonZero`) to the list in `Standard` type documentation
* implemented `Distribution<maskNxM> for Standard`, behaves like `rng.gen::<bool>`
* ~~implemented `Distribution<mask64xM> for Bernoulli`, each lane uses the same prob~~

The trait system allows impls to adapt to future SIMD types and doesn't clutter the rendered documentation. Using it in more places would mean some code duplication though.

`__m512i` requires nightly Rust's `stdsimd` feature at the moment so I put it under the `simd_support` feature but `nightly` might make more sense.

TheIronBorn commented:
Turns out we don't have any SIMD benchmarks

dhardy reviewed (commented):
Good riddance, fragile `packed_simd`! Lets hope std SIMD is better.

This isn't a full review. Thanks @TheIronBorn.

TheIronBorn commented on benches/misc.rs:
```diff
@@ -75,7 +75,7 @@ fn misc_bernoulli_const(b: &mut Bencher) {
         let mut accum = true;
         for _ in 0..crate::RAND_BENCH_N {
-            accum ^= rng.sample(d);
+            accum ^= rng.sample::<bool, _>(d);
```
oh right I didn't consider that. That is rather nasty

TheIronBorn commented:
I'm agreed that introducing seemingly unpredictable type inference is bad.

An extra trait or something would also make documentation easier, we could use `#[cfg_attr(doc_cfg, doc(cfg(feature = "simd_support")))]` and other stuff as well.

The Bernoulli stuff should perhaps wait for #1227

TheIronBorn commented:
Removed the Bernoulli stuff and squashed some commits.

Ran some benchmarks for 128/256 bit vectors and they're mostly the same except for small optimizations packed_simd had but std::simd isn't prioritizing. Maybe LLVM might optimize it better in the future.

dhardy commented on src/distributions/integer.rs:
```diff
@@ -109,53 +114,63 @@ impl_nzint!(NonZeroU64, NonZeroU64::new);
-                    let b_ptr = &mut *(ptr as *mut $ty as *mut [u8; $bits/8]);
-                    rng.fill_bytes(b_ptr);
+                // TODO: impl this generically once const generics are robust enough
+                let mut vec: Simd<u8, { mem::size_of::<$ty>() }> = Default::default();
+                rng.fill_bytes(vec.as_mut_array());
+                // NOTE: replace with `to_le` if added to core::simd
+                #[cfg(not(target_endian = "little"))]
+                {
+                    vec = vec.reverse();
                 }
-                vec.to_le()
+                // SAFETY: we know u8xN and $ty have the same size
+                unsafe { mem::transmute_copy(&vec) }
```
If you use `Fill` (with the correct element type) it will do the Endianness conversion here. `reverse` is not the correct conversion IIUC.

TheIronBorn replied:
Oh nice that'll also make it more generic

dhardy commented on src/distributions/other.rs:
```diff
@@ -145,6 +147,37 @@ impl Distribution<bool> for Standard {
+    Standard: Distribution<Simd<T, LANES>>,
+{
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Mask<T, LANES> {
+        rng.gen().lanes_lt(Simd::default())
+    }
```
Doesn't this compute to `false` in all lanes?

And, regarding the proceeding doc, ... doesn't this make the `Mark` type inappropriate if normally only one bit from each lane is used? Ugh, guess we can't fix that.

BTW one can assume that a random boolean is `true` with 50% probability, but should one actually assume anything about what a random `Mask` is?

Maybe instead we should provide a higher level API around `select` etc... though given that SIMD is a little niche, maybe just a few recipes in the book would be enough.

dhardy replied:
No specific use-case; I was just going by what you mentioned (select). Probably this is unnecessarily to get into however.

dhardy commented:
> Realized that the `nightly` feature is documented to only provide optimizations, not extra API stuff so putting `__m512` under `simd_support` probably makes the most sense

Any flags depending on `nightly` are unstable so we don't have to worry too much. I would guess if other SIMD parts get stabilized in libstd first, then `__m512` would eventually end up under another feature flag.. or `simd_support` would end up only adding the non-stabilized SIMD features.

TheIronBorn commented:
Anything left to review here?

TheIronBorn commented:
Oh there's a weird new failure. I don't understand why it seems to be enabling `simd_support` with just `nightly`

dhardy commented on src/distributions/other.rs:
````diff
@@ -145,6 +147,39 @@ impl Distribution<bool> for Standard {
 
+/// Requires nightly Rust and the [`simd_support`] feature
+///
+/// Note that on some hardware like x86/64 mask operations like [`_mm_blendv_epi8`]
+/// only care about a single bit. This means that you could use uniform random bits
+/// directly:
+///
+/// ```ignore
+/// // this may be faster...
+/// let x = unsafe { _mm_blendv_epi8(a.into(), b.into(), rng.gen::<__m128i>()) };
+///
+/// // ...than this
+/// let x = rng.gen::<mask8x16>().select(b, a);
+/// ```
+///
+/// Since most bits are unused you could also generate only as many bits as you need.
+///
+/// [`_mm_blendv_epi8`]: https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm_blendv_epi8&ig_expand=514/
+/// [`simd_support`]: https://github.com/rust-random/rand#crate-features
+#[cfg(feature = "simd_support")]
+impl<T, const LANES: usize> Distribution<Mask<T, LANES>> for Standard
+where
+    T: MaskElement + PartialOrd + SimdElement<Mask = T> + Default,
+    LaneCount<LANES>: SupportedLaneCount,
+    Standard: Distribution<Simd<T, LANES>>,
+{
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Mask<T, LANES> {
+        // `MaskElement` must be a signed integer, so this is equivalent
+        // to the scalar `i32 < 0` method
+        rng.gen().lanes_lt(Simd::default())
+    }
+}
````
From what I understand this is correct, but not necessarily the most efficient: only the high bit of each lane generated by `rng.gen()` is used. I'm not sure if this is actually worth worrying about however.

I'm also just a little surprised it type-checks: it relies on only one type supporting `lanes_lt(..) -> Mask<T, LANES>`. Better to clarify with `rng.gen::<T>().lates_lt(T::default())`?

TheIronBorn replied:
The inefficiency can actually be worse in scalar. `mask8x16` wastes 7 bits per lane while `bool` wastes 31, `mask64x2` though wastes 63. I'll look into more efficient methods but might just point readers to it in the docs

dhardy reviewed (approved):
But the above comments aren't blockers, so I'll also approve this PR.

TheIronBorn commented on src/distributions/integer.rs:
```diff
@@ -109,53 +114,54 @@ impl_nzint!(NonZeroU64, NonZeroU64::new);
-#[cfg(feature = "simd_support")]
-macro_rules! simd_impl {
-    ($(($intrinsic:ident, $vec:ty),)+) => {$(
+macro_rules! intrinsic_impl {
```
Agreed

TheIronBorn commented:
Ah, the tests are failing because the stdsimd API changed in a recent nightly. Shame, I hoped stdsimd would be more stable, though I suppose it's a different kind of instability.

We could add some compilation conditions or wait until the changes reach stable.

dhardy commented:
I'm sure you know more than I do about the state of `std::simd` so I'll let you decide on that. Our status quo is also broken.

TheIronBorn commented:
we could simply merge the packed_simd fix while we wait

dhardy commented:
Can do. The only thing against is that your migrations here will likely end up stale. Any idea on the time frame for stability or how much churn is likely to `std::simd`?

TheIronBorn commented:
I don't know what I was thinking earlier but since std::simd is only available on nightly we can just fix and merge right now

TheIronBorn commented:
Unfortunately there's no roadmap for stdsimd but there is likely to be at least a little churn. Though we aren't using too many features so I think we won't be affected much. 
And at least it's more likely to be API instability than forcing us to rollback as often as packed_simd

TheIronBorn commented:
The only remaining failure is crossbeam
